### PR TITLE
Add webhooks: `block.created`, `mute.created`, `follow.created`

### PR DESCRIPTION
**Pull request by CSDUMMI**
This PR adds Webhooks for when certain user interactions take place on an instance. 

The permissions required for these webhooks are based on my best guess and I'm open to changing them. 

Example bodies sent to a Webhook Endpoint (for future documentation and app developers reference):

#### `block.created`
```json
{
    "event": "block.created",
    "created_at": "2024-01-15T19:22:28.134Z",
    "object": {
        "id": 1,
        "created_at": "2024-01-15T19:22:28.025Z",
        "updated_at": "2024-01-15T19:22:28.025Z",
        "account_id": 110482205265885150,
        "target_account_id": 111484398945680759,
        "uri": "https://<mastodon.example>/68e9cf81-ec5e-40ab-a775-1e510c83fcea"
    }
}
```

#### `follow.created`
```json
{
    "event": "follow.created",
    "created_at": "2024-01-15T19:22:46.080Z",
    "object": {
        "id": 1,
        "created_at": "2024-01-15T19:22:46.014Z",
        "updated_at": "2024-01-15T19:22:46.014Z",
        "account_id": 110482205265885150,
        "target_account_id": 111484398945680759,
        "show_reblogs": true,
        "uri": "https://<mastodon.example>/8cbcffcc-14a2-4f1f-baa9-15cb42af78a2",
        "notify": false,
        "languages": null
    }
}
```

#### `mute.created`
```json
{
    "event": "mute.created",
    "created_at": "2024-01-15T19:22:59.911Z",
    "object": {
        "id": 1,
        "created_at": "2024-01-15T19:22:59.899Z",
        "updated_at": "2024-01-15T19:22:59.899Z",
        "hide_notifications": true,
        "account_id": 110482205265885150,
        "target_account_id": 111484398945680759,
        "expires_at": null
    }
}
```

I'm working on creating Webhooks for the inverse of these actions: `block.removed`, `follow.removed` and `mute.removed` but am separating these out as they work more like the other already implemented Webhooks.

### CODE REVIEW

File: app/services/follow_service.rb [hunk 1]
```diff
@@ -39,7 +39,9 @@ def call(source_account, target_account, options = {})
     if (@target_account.locked? && !@options[:bypass_locked]) || @source_account.silenced? || @target_account.activitypub?
       request_follow!
     elsif @target_account.local?
-      direct_follow!
+      follow = direct_follow!
+      trigger_webhooks follow.id
```
Comment (mjankowski) [inline app/services/follow_service.rb:43]: Looks like `direct_follow!` already has some calls to other workers ... maybe just add there (avoids the local var and explicit return up here)

Comment (CSDUMMI) [reply app/services/follow_service.rb:43]: Didn't see that. This is better indeed. Also makes the `trigger_webhooks` function obsolete.

**Comment by ThisIsMissEm** [comment]
Could you explain in the PR description why you need these webhooks? I'm not sure I understand when these would be useful?

**Comment by ArneBab** [comment]
@ThisIsMissEm these webhooks enable connecting external protection against spam and harrassment.

As concrete exampple they make it possible to connect the NLnet-supported wispwot project to get propagating visibility and collaborative blocking which makes spamming and harrassment scale worse than blocking spammers:
https://nlnet.nl/project/Wispwot/

Simplified: people your friends follow are visible, but if enough of your friends block an account, that account becomes invisible for you, too. It’s the intermediate solution between blocking someone personally and blocking someone on a whole instance.

(the method has proven itself in over 10 years of operation in fully pseudonymous environments where there cannot be personal consequences beyond losing visibility, and it actually works in keeping conversations civilized: https://www.draketo.de/english/freenet/friendly-communication-with-anonymity )

But these changes here do not just enable this single tool. They are built to ease the development of tools for moderation, so every instance administrators can experiment with different moderation tools.

**Comment by ThisIsMissEm** [comment]
@ArneBab thanks for that description and explanation, though I cannot help but wonder if this would be better done at user level instead of instance level? Webhooks are at present instance level, meaning you receive everything for all users, and an admin has to opt-in to using your service.

Doing this at user-level would give an experience more like that of BlockParty or BlockTogether and users wouldn't need to rely on their admin consenting to send all data to a potentially third-party system. 

I'm aware wispwot can be run locally, but this would mean that you're not gaining from network effects on smaller instances, since you'd only have data in your server from your server.

I do think there is a use-case to allow OAuth Applications to register a webhook tied to them, based on their scopes already available, which could provide this data more real-time than polling registered users' blocks & mutes.

But I don't think webhooks for blocks, mutes and follows at instance level is a good idea, partly due to needing an admin to set them up, and partly due to how this data could be misused.

From personal experience with IFTAS CCS, which consumes `status.*` webhook events, we've had a significant amount of feedback of admins being unsure if they wish to grant us access to all that data, despite data processing agreements and measures taken to keep that data safe; I suspect you may face the same

**Comment by ArneBab** [comment]
@ThisIsMissEm wispwot is designed to run locally. The follow-, mute- and block-data on the local instance should provide enough information about relevant profiles on other instances to limit the impact of actions from profiles on those instances: once remote people start to disrupt communications, people on the local instance will start to mute and block them, so the remote disrupting profiles will lose visibility for most local profiles — except the ones who explicitly follow them. This also means that no full consensus about blocking is required: efficient self-protection is possible without disrupting communications of people who want to interact with those remote profiles.

This should lessen the load on the moderation team. And that’s direly needed — I know of one instance where many mods are already deeply jaded because they have to create instance block after instance block to protect their users (there are many people just throwing hate messages into discussions they have no part in).

The advantage of the webhooks here is that they can also be used to inform other anti-spam tools: they are not tied to wispwot (actually wispwot was changed to support these webhooks to minimize changes to Mastodon).

In short: wispwot is built to run in the same environment as the instance, so no data flows outside, and using it gives instances that are not just an instance-of-one a basic level of purely user-controlled moderation that does not require interaction by the mods.

**Comment by mjankowski** [comment]
I think a rebase here would get the current CI config, and probably remove the annotations from the diff.

Separately - maybe add some basic specs around the worker changes?

**Comment by ThisIsMissEm** [comment]
I'm not sure this makes sense to land.

**Comment by mjankowski** [comment]
> I'm not sure this makes sense to land.

Cool.

> wispwot is designed to run locally. The follow-, mute- and block-data on the local instance should provide enough information [...]

I'm not a wispwot expert - can you say more about how it runs locally?

**Comment by mjankowski** [comment]
Closing on assumption that zero feedback from core team in ~1 year implies a lack of desire to merge this; and previous concerns about mixing the user/instance level on this from above.

If you want to push forward, can you add some more detail about the wispwot use case either here or on new issue or whatever works (and/or rebase, reopen, address feedback, etc...)

**Comment by ArneBab** [comment]
I’m sorry for not answering: I missed your question.

Wispwot builds a communication graph in which following creates a positive edge between follower and followed, muting removes that edge and adds a note to ignore the ignored, and blocking creates a negative edge.

Positive edges are followed and the values for followed and blocked accounts (positive and negative) are aggregated into a score which decides whether to show messages by the accounts. The positive and negative values are weighted by distance from your account, so follows and blocks by people you follow have more weight than those by accounts three or four steps removed.

This way softer moderation decisions can be computed quickly from individual actions without putting load on moderators (who are already a bottleneck on many instances).

**Comment by ArneBab** [comment]
@mjankowski did you see my reply two months ago?

**Comment by mjankowski** [comment]
Yes, thanks!
